archeryutils/handicaps/handicap_scheme.py

```diff
@@ -0,0 +1,764 @@
+        ``int_prec=True``:
+
+        >>> agb_scheme.handicap_from_score(999, wa_outdoor.wa1440_90), int_prec=True)
+        44.0
+
+        """
+        # Check we have a valid score
+        max_score = rnd.max_score()
+        if score > max_score:
+            raise ValueError(
+                f"The score of {score} provided is greater than the maximum of {max_score} "
+                f"for a {rnd.name}."
+            )
+        if score <= 0.0:
+            raise ValueError(
+                f"The score of {score} provided is less than or equal to zero so cannot "
+                "have a handicap."
+            )
+
+        if score == max_score:
+            # Deal with max score before root finding
+            return self._get_max_score_handicap(rnd, arw_d, int_prec)
+
+        handicap = self._rootfind_score_handicap(score, rnd, arw_d=arw_d)
+
+        # Force integer precision if required.
+        if int_prec:
+            if self.desc_scale:
+                handicap = np.ceil(handicap)
+            else:
+                handicap = np.floor(handicap)
+
+            sc_int = self.score_for_round(rnd, handicap, arw_d, rounded_score=True)
+
+            # Check that you can't get the same score from a larger handicap when
+            # working in integers
+            min_h_flag = False
+            if self.desc_scale:
+                hstep = 1.0
+            else:
+                hstep = -1.0
+            while not min_h_flag:
+                handicap += hstep
+                sc_int = self.score_for_round(rnd, handicap, arw_d, rounded_score=True)
+                if sc_int < score:
+                    handicap -= hstep  # undo the iteration that caused flag to raise
+                    min_h_flag = True
+
+        return handicap
+
+    def _get_max_score_handicap(
+        self,
+        rnd: rounds.Round,
+        arw_d: Optional[float] = None,
+        int_prec: bool = False,
+    ) -> Union[int, float]:
+        """Get handicap for maximum score on a round.
+
+        Start high and drop down until no longer rounding to max score.
+        i.e. >= max_score - 1.0 for ceil() rounding, and >= max_score - 0.5 for around().
+
+        Parameters
+        ----------
+        rnd : rounds.Round
+            round being used
+        arw_d : float or None, default=None
+            user-specified arrow diameter in [metres]
+        int_prec : bool, default=False
+            display results as integers?
+
+        Returns
+        -------
+        handicap : int or float
+            Handicap for maximum score. Has type int if int_prec is True, else float.
+
+        Warns
+        ------
+        UserWarning
+            If called with int_prec=False as value limited by the numerical scheme delta.
+
+        """
+        max_score = rnd.max_score()
+
+        if self.desc_scale:
+            handicap = -75.0
+            delta_hc = 0.01
+        else:
+            handicap = 175.0
+            delta_hc = -0.01
```

Comment:
Dispatching on desc scale is coupling the base method to the specifics of the AGB and AA systems. As with the rouding limit I think its better to make these look up named variables on the schemes themselves, perhaps something like:
```
# In AGB subclasses
absolute_minimum_handicap=-75.0
desc_scale = True

# in AA subclasses
absolute_minimum_handicap=175.0
desc_scale = False

# in this method
handicap = self.absolute_minimum_handicap
delta_hc = 0.01 * (-1 if not self.desc_scale else 1)
```

Obviously naming etc to taste but I think it makes sense to decouple.

Reply:
Do you feel https://github.com/jatkinson1000/archeryutils/pull/78/commits/6ce0f78fd8a4ee47e0946cb96d199f30a25be49e resolves this?
I think it can still be set on `desc_scale`, as the direction of the search will determine whether we are searching from low to high or vice versa, but I have removed the implicit dependence on specific schemes to make the start point set from a handicap scheme attribute (`scale_bounds`).

Reply:
I also think the implemented approach is slightly clearer as it indicated whether we are searching from the max or min bound in each case, rather than having a variable `min_handicap` that takes a maximal numerical value.

Reply:
Yeah, that does look better. If I'm being picky I'd still prefer to see as little data set as possible in the base class, rather just the type declarations on the variables and forcing all subclasses to set their data explicitly (makes it **much** easier to track down that data too if its all in one place), so I think it would be better for `desc_scale` and `scale_bounds` (and the `max_score_rounding_lim`) to be set explicity in HandicapAGB and HandicapAGBold. Even with the helpful comment above the variable declarations in the base, this relys a bit too much on implicit inheritance when the data could be explicitly defined where its applied.

Reply:
To be set as an instance variable in the base class (required to satisfy the code that uses them in the base class methods) they have to be declared and given a value (I don't think we can have uninitialised variables in python).

I do get what you're saying however, so options might be:
- Leave as is, but explicitly overwrite every variable in each classification subclass to reinforce.
- Initialise with junk (`0.0`, 'True' etc.) and explicitly overwrite every variable in each classification subclass.
  but I'm not keen on placing a junk default there in case.
- Outfit the code using [pydantic](https://docs.pydantic.dev/latest/) which would allow implementing as described (plus perhaps some [other bonus features](https://www.youtube.com/watch?v=C-GCX9Ftzko)).

Reply:
Had a go at the pydantic approach - What do you think to #79?

Reply:
Ah, thats annoying, my comment here never posted as it got lost in a pending review!!

I was trying to say theres an easier way, you dont *need* pydantic for this:

```python
class HandicapScheme(ABC):
    desc_scale: bool    
    scale_bounds: list[float]
    max_score_rounding_lim: float
```

Just declare the variables with type annotations (dataclass style); mypy is happy, no junk variables, can then overwrite in every subclass

Reply:
😅 Having been through it I do like the neatness of the pydantic approach in #79 which cleans up some things such as not having an `__init__` and neat addition of other scheme-specific parameters.
The extra checks it brings are also nice from a security point of view, but could be argued as overkill given much of this is internal rather than user-facing code.

Reply:
I do agree the code looks better after the diff there, but at the cost of a bit of technical debt. I feel the need to figure out what magic things pydantic is or isn't doing if I want to change anything in that part of the code, and I'm not personally familiar with it. I think we can get it nearly as clean without it so personally I wouldn't bother with the extra dependency. But also its pretty small scale on a part of the code I'm unlikely to touch so I wouldn't straight up oppose it.

Reply:
> ```python
> class HandicapScheme(ABC):
>     desc_scale: bool    
>     scale_bounds: list[float]
>     max_score_rounding_lim: float
> ```
> 
> Just declare the variables with type annotations (dataclass style); mypy is happy, no junk variables, can then overwrite in every subclass

In this case they're being declared as class attributes rather than instance attributes, however, which is what @LiamPattinson raised as an issue in https://github.com/jatkinson1000/archeryutils/pull/78#discussion_r1500868639. Whilst we could then overwrite as class attributes in the subclasses this perhaps smells a bit off/is a bit obscure - a value that has to be set specifically for every subclass.
After sleeping on it pydantic may also help with type validation for `Target` and `Round`, allowing failure at initialisation rather than in a deeper routine at runtime...

Reply:
I don't use MyPy very often, and I tend to use type annotations just as in-code documentation, so I'm not totally clued up on this stuff, but I think that the following doesn't declare class variables:

```python
class HandicapScheme(ABC):
    desc_scale: bool    
    scale_bounds: list[float]
    max_score_rounding_lim: float
```

Since you don't assign to them, no class variables are created, and only static type checkers pay attention to those lines. They'll be interpreted as instance variables that will be set at some point, and to instead get an annotation for a class variable you need to write `myclassvar: ClassVar[int]`.

Reply:
Huh, OK.
That said, I'm not super keen on this as:
- It doesn't readily seem super clear whether they are class or instance variables.
- It doesn't really create items which are used as part of the class methods.

Reply:
This was in my original comment that got swallowed by the review, I didn't rewrite it but you can also totally declare the variable types in the base class init if that is more to your tastes:

```
class HandicapScheme(ABC):
    def __init__(self, *args):
        self.desc_scale: bool    
        self.scale_bounds: list[float]
        self.max_score_rounding_lim: float
```

I have a diff that literally just changes this, declares types in the base init, declares all relavent variables in the subclass inits (and removes the redundant type annotations there), passes all tests and satisfies mypy too.
```
git diff --stat
 archeryutils/handicaps/handicap_scheme.py     | 10 +++++-----
 archeryutils/handicaps/handicap_scheme_aa.py  |  6 ++++--
 archeryutils/handicaps/handicap_scheme_agb.py | 14 +++++++++++++-
 3 files changed, 22 insertions(+), 8 deletions(-)
```

Reply:
OK, this should be implemented in https://github.com/jatkinson1000/archeryutils/pull/78/commits/d7ac0ec01a1906ec077b95576f6e29eb1782488d Are we all now happy with this implementation?

Reply:
Looking at the last patch all good from me